src/java.base/share/classes/jdk/internal/foreign/abi/Specializer.java

```diff
@@ -0,0 +1,854 @@
+            Path dumpPath = Path.of(DUMP_CLASSES_DIR).resolve(fileName);
+            try {
+                Files.createDirectories(dumpPath.getParent());
+                Files.write(dumpPath, bytes);
+            } catch (IOException e) {
+                throw new InternalError(e);
+            }
+        }
+
+        if (PERFORM_VERIFICATION) {
+            boolean printResults = false; // only print in case of exception
+            CheckClassAdapter.verify(new ClassReader(bytes), null, printResults, new PrintWriter(System.err));
+        }
+
+        try {
+            // We must initialize the class since the upcall stubs don't have a clinit barrier, and the slow
+            // path in the c2i adapter we end up calling can not handle the particular code shape where the
+            // caller is an optimized upcall stub.
+            boolean initialize = callingSequence.forUpcall();
+            MethodHandles.Lookup lookup = MethodHandles.lookup().defineHiddenClassWithClassData(bytes, leafHandle, initialize);
+            return lookup.findStatic(lookup.lookupClass(), METHOD_NAME, callerMethodType);
+        } catch (IllegalAccessException | NoSuchMethodException e) {
+            throw new InternalError("Should not happen", e);
+        }
+    }
+
+    private static String escapeForFileName(String str) {
+        StringBuilder sb = new StringBuilder(str.length());
+        for (int i = 0; i < str.length(); i++) {
+            char c = str.charAt(i);
+            sb.append(switch (c) {
+                case ' ' -> '_';
+                case '[', '<' -> '{';
+                case ']', '>' -> '}';
+                case '/', '\\', ':', '*', '?', '"', '|' -> '!'; // illegal in Windows file names.
+                default -> c;
+            });
+        }
+        return sb.toString();
+    }
+
+    private void specialize() {
+        // map of parameter indexes to local var table slots
+        paramIndex2ParamSlot = new int[callerMethodType.parameterCount()];
+        for (int i = 0; i < callerMethodType.parameterCount(); i++) {
+            paramIndex2ParamSlot[i] = newLocal(callerMethodType.parameterType(i));
+        }
+
+        // slots that store the output arguments (passed to the leaf handle)
+        leafArgSlots = new int[leafType.parameterCount()];
+        for (int i = 0; i < leafType.parameterCount(); i++) {
+            leafArgSlots[i] = newLocal(leafType.parameterType(i));
+        }
+
+        // allocator passed to us for allocating the return MS (downcalls only)
+        if (callingSequence.forDowncall()) {
+            RETURN_ALLOCATOR_IDX = 0; // first param
+        }
+
+        // create a Binding.Context for this call
+        if (callingSequence.allocationSize() != 0) {
+            emitConst(callingSequence.allocationSize());
+            mv.visitMethodInsn(INVOKESTATIC, BINDING_CONTEXT_INTRN, OF_BOUNDED_ALLOCATOR_NAME, OF_BOUNDED_ALLOCATOR_DESC, false);
```

Comment:
In this, and other instructions, I wonder if it would be more readable to have helper routine, and rely on class literal/strings more e.g. instead of:
```
            mv.visitMethodInsn(INVOKESTATIC, BINDING_CONTEXT_INTRN, OF_BOUNDED_ALLOCATOR_NAME, OF_BOUNDED_ALLOCATOR_DESC, false);
```
something like:

```
emitInvokeStatic(Binding.Context.class, "ofBoundedAllocator", OF_BOUNDED_ALLOCATOR_DESC);
```

Reply:
That's a good idea. I think we can use DirectMethodHandleDesc to describe the method we want to call